File: mirror/tier2-ruby/showcase/sec_injection.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Security
# Module: sec_injection.rb
# Purpose: Demonstrate injection prevention

def escape_sql(value)
  return "" unless value.is_a?(String)
  
  value
    .gsub("'", "''")
    .gsub('"', '""')
    .gsub("\\", "\\\\")
end

def validate_query(query)
  return false unless query.is_a?(String)
  
  forbidden = ["DROP", "DELETE", "UPDATE", "INSERT", "ALTER"]
  upper_query = query.upcase
  
  forbidden.none? { |keyword| upper_query.include?(keyword) }
end

def sanitize_html(content)
  return "" unless content.is_a?(String)
  
  content
    .gsub(/<script[^>]*>.*?<\/script>/i, '')
    .gsub(/on\w+\s*=/i, '')
end

result = {
  escaped_quote: escape_sql("user's name"),
  valid_select: validate_query("SELECT * FROM users"),
  invalid_drop: validate_query("DROP TABLE users"),
  invalid_delete: validate_query("DELETE FROM data"),
  sanitized_html: sanitize_html("hello<script>alert(1)</script>"),
  injection_prevention_active: true,
}

puts result.inspect

File: mirror/tier2-python/showcase/sec_injection.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Security
Module: sec_injection.py
Purpose: Demonstrate injection prevention
"""

import re


def escape_sql(value: str) -> str:
    """Escape SQL special characters"""
    if not isinstance(value, str):
        return ""
    
    escapes = {
        "'": "''",
        '"': '""',
        "\\": "\\\\",
    }
    
    result = value
    for char, escaped in escapes.items():
        result = result.replace(char, escaped)
    
    return result


def validate_query(query: str) -> bool:
    """Validate query for dangerous operations"""
    if not isinstance(query, str):
        return False
    
    forbidden = ["DROP", "DELETE", "UPDATE", "INSERT", "ALTER"]
    upper_query = query.upper()
    
    for keyword in forbidden:
        if keyword in upper_query:
            return False
    
    return True


def sanitize_html(content: str) -> str:
    """Sanitize HTML content"""
    if not isinstance(content, str):
        return ""
    
    # Remove script tags and dangerous attributes
    content = re.sub(r'<script[^>]*>.*?</script>', '', content, flags=re.IGNORECASE)
    content = re.sub(r'on\w+\s*=', '', content, flags=re.IGNORECASE)
    
    return content


result = {
    "escaped_quote": escape_sql("user's name"),
    "valid_select": validate_query("SELECT * FROM users"),
    "invalid_drop": validate_query("DROP TABLE users"),
    "invalid_delete": validate_query("DELETE FROM data"),
    "sanitized_html": sanitize_html("hello<script>alert(1)</script>"),
    "injection_prevention_active": True,
}

if __name__ == "__main__":
    print(result)
